### PR TITLE
[IRGen] Call objc_direct methods correctly

### PR DESCRIPTION
**Pull request by ellishg**
When we encounter an Objective-C method with the `objc_direct` attribute we don't want to use `objc_msgSend()`. Instead we call the method directly.

### CODE REVIEW

File: include/swift/AST/Decl.h
Comment (CodaFi) [inline include/swift/AST/Decl.h:5811]: I don't think we need to leak details of the calling convention all the way up into the surface language's AST. We can pull this information out of the corresponding clang node at [SIL/IR]Gen time.

File: test/IRGen/Inputs/objc_direct.h
Comment (CodaFi) [inline test/IRGen/Inputs/objc_direct.h:3]: Please also add a test for `__attribute__((objc_direct_members))`, and categories with direct methods.

File: lib/IRGen/IRGenSIL.cpp
Comment (CodaFi) [inline lib/IRGen/IRGenSIL.cpp:2274]: Perhaps it's better to resolve this with a new `SILFunctionTypeRepresentation::ObjCDirectMethod`. @rjmccall, what do you think?

Comment (ellishg) [reply lib/IRGen/IRGenSIL.cpp:2274]: Let me know if you'd like to have a `SILFunctionTypeRepresnetation::ObjCDirectMethod` and I can add it.

**Comment by ellishg** [comment]
I've resolved most of the comments. Can I get a second review? @CodaFi @rjmccall

File: lib/SILGen/SILGenApply.cpp
Comment (slavapestov) [inline lib/SILGen/SILGenApply.cpp:1061]: Can you add a test for Swift protocol conformances where the requirement is satisfied by an `objc_direct` witness? Emission of the call to the witness from the body of a witness thunk doesn't go through SILGenApply. I think we might need to insert a similar check in getWitnessDispatchKind() or thereabouts.

File: test/IRGen/objc_direct.swift
Comment (slavapestov) [inline test/IRGen/objc_direct.swift:1]: Please also add a SILGen test that uses FileCheck to verify that the call is lowered as a function_ref and not an objc_method.

More ideas for things to test:
- Protocol conformances, as above
- Calling an objc_direct method on AnyObject -- should probably disallow this?
- Can (methods that get imported as) subscripts be objc_direct?

File: test/IRGen/objc_direct.swift
Comment (slavapestov) [inline test/IRGen/objc_direct.swift:5]: I think it would be more consistent to split this up into an IRGen test and an execution test in test/Interpreter. The execution test might need to be guarded by an availability check if objc_direct methods don't backward deploy (I don't know if they do or not).

**Review by slavapestov — COMMENTED** [review]
This looks great!

File: lib/IRGen/IRGenSIL.cpp
Comment (slavapestov) [reply lib/IRGen/IRGenSIL.cpp:2274]: Instead of adding a new SILFunctionType representation, I would suggest a new CalleeKind instead.

File: lib/ClangImporter/ImportDecl.cpp
Comment (rjmccall) [inline lib/ClangImporter/ImportDecl.cpp:7684]: Hmm.  When do we import an ObjCMethodDecl as something besides an AbstractFunctionDecl?  If this triggers for something like properties, we really should support it rather than making it unavailable.

File: lib/IRGen/IRGenSIL.cpp
Comment (rjmccall) [reply lib/IRGen/IRGenSIL.cpp:2274]: Yeah, I agree with Slava.

File: lib/IRGen/GenDecl.cpp
Comment (rjmccall) [inline lib/IRGen/GenDecl.cpp:2752]: It's unfortunate that we can't just ask Clang for the declaration here.

File: test/IRGen/objc_direct.swift
Comment (rjmccall) [reply test/IRGen/objc_direct.swift:1]: I think we also need typechecker tests that we don't allow overriding, either handle or disallow super calls, and so on.  Probably the override part means importing the method as final.

File: lib/ClangImporter/ImportDecl.cpp
Comment (slavapestov) [reply lib/ClangImporter/ImportDecl.cpp:7684]: I suspect for properties we should still be mapping the ObjCMethodDecl over to the AccessorDecl on the VarDecl and not the VarDecl itself, but I could be wrong.

File: lib/SILGen/SILGenApply.cpp
Comment (ellishg) [reply lib/SILGen/SILGenApply.cpp:1061]: I ran into this error https://github.com/llvm/llvm-project/blob/20d704a75ed51c7a9a155aa3978d0c02671c3f69/clang/test/SemaObjC/method-direct.m#L64 while trying to add this test. Since ObjC doesn't allow `objc_direct` on protocols, doesn't that make this code pattern impossible?

Comment (theblixguy) [reply lib/SILGen/SILGenApply.cpp:1061]: I think you need a Swift protocol and ObjC class that implements that protocol and where the witness is annotated with `__attribute__((objc_direct))`.

File: lib/ClangImporter/ImportDecl.cpp
Comment (ellishg) [reply lib/ClangImporter/ImportDecl.cpp:7684]: For properties and subscripts, they don't seem to hit this error and I haven't found anything that does. Should we remove this? I kept in the diagnostic in case there was an edge case I didn't know about.

File: lib/IRGen/IRGenSIL.cpp
Comment (ellishg) [reply lib/IRGen/IRGenSIL.cpp]: I've added `getObjCDirectMethodCallee()` to generate the callee here.

File: test/IRGen/objc_direct.swift
Comment (ellishg) [reply test/IRGen/objc_direct.swift]: Thanks for the suggestions! The tests have been refactored into `SILGen`, `IRGen`, `Interpreter`, and `ClangImporter` (disallow `objc_direct` on `AnyObject`). I've also added subscript methods to the tests.

File: lib/SILGen/SILGenApply.cpp
Comment (ellishg) [reply lib/SILGen/SILGenApply.cpp:1061]: It shouldn't matter whether the original protocol is in Swift for ObjC. When clang runs, it will see an ObjC protocol and disallow any `objc_direct` methods on that protocol.

File: test/IRGen/objc_direct.swift
Comment (ellishg) [reply test/IRGen/objc_direct.swift]: @rjmccall I've added a test in `Interpreter/objc_direct.swift` and it appears that overriding `objc_direct` methods works fine.
https://github.com/apple/swift/pull/33349/files#diff-be1eb3d9c772960853ed4997fcf00754R43

File: lib/SILGen/SILGenApply.cpp
Comment (slavapestov) [reply lib/SILGen/SILGenApply.cpp:1061]: Both the protocol and the conformance should be defined in Swift. Eg,

```
protocol MyProto {
  func myMethod()
}

extension MyObjCClass : MyProto {}
```

Then MyObjCClass would have an `objc_direct` method named `myMethod()`.

File: lib/ClangImporter/ImportDecl.cpp
Comment (slavapestov) [reply lib/ClangImporter/ImportDecl.cpp:7684]: I would suggest changing the `dyn_cast` to a `cast` to make it assert if we ever hit this case.

File: test/IRGen/objc_direct.swift
Comment (rjmccall) [reply test/IRGen/objc_direct.swift]: What you're testing is whether an imported direct method can be used as a protocol witness.  That's good, but it's not what I meant; I meant a class method override.  You should check that you can't define a subclass of an ObjC class that overrides a direct method.

File: lib/IRGen/IRGenSIL.cpp
Comment (rjmccall) [inline lib/IRGen/IRGenSIL.cpp:2272]: We have to do something extra on the caller side when making a direct call of a class method, right?

File: lib/SIL/IR/SILDeclRef.cpp
Comment (rjmccall) [inline lib/SIL/IR/SILDeclRef.cpp:701]: Okay, no.  We should have a way to get this declaration that doesn't duplicate code across projects like this.  If that means that the mangling logic for direct methods in Clang needs to be moved into clang/AST, so be it.

File: lib/AST/NameLookup.cpp
Comment (brentdax) [reply lib/AST/NameLookup.cpp:1822]: This is effectively a choice between emitting a generic and slightly incorrect "no method found"-type diagnostic, and emitting a tailored "direct methods cannot be called through AnyObject dispatch"-type diagnostic.

I think it'd be better to emit a tailored diagnostic, but it could be left out if that would be very difficult, or it could be emitted based on a separate lookup performed while diagnosing the error. Please make sure you include a test case where one class has a direct method and another has a normal method by the same name—we'll want to make sure that the direct method doesn't stop you from using the normal one through AnyObject dispatch.

File: lib/IRGen/IRGenSIL.cpp
Comment (ellishg) [reply lib/IRGen/IRGenSIL.cpp:2272]: What else is needed on the caller side? I have objc_direct class method test cases (`directClassMethod`, `directClassMethod2`) that seem to be working.

File: lib/AST/NameLookup.cpp
Comment (ellishg) [reply lib/AST/NameLookup.cpp:1822]: I do agree that this is a very unhelpful error message for developers who might be surprised to see a "no method found"-type message after adding an `objc_direct` attribute. I'm not very familiar with this part of the codebase, where would we put the lookup to diagnose the error?

And thanks for the test case, I'll add that.

File: lib/IRGen/IRGenSIL.cpp
Comment (rjmccall) [reply lib/IRGen/IRGenSIL.cpp:2272]: Well, in general you have to make sure the class is realized before you can call direct class methods.  But in fact it looks like Swift does this explicitly before calling *any* class method, which is unfortunate — but it means there's nothing to do here except make sure we test that pattern, so that if Swift ever gets smarter here we can avoid some work.

File: lib/SIL/IR/SILDeclRef.cpp
Comment (ellishg) [reply lib/SIL/IR/SILDeclRef.cpp:701]: I've submitted https://reviews.llvm.org/D88329 which makes `MangleContext::mangleObjCMethodNameWithoutSize()` available to use here.

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

File: lib/IRGen/IRGenSIL.cpp
Comment (ellishg) [reply lib/IRGen/IRGenSIL.cpp:2272]: I think this makes sense, but I don't know how to create this test. Would it be something like calling a class method that has an interface but no implementation?

Comment (rjmccall) [reply lib/IRGen/IRGenSIL.cpp:2272]: You can just do test the IRGen output for a call to a direct class method that looks something like `TheClass.directClassMethod()`.

Comment (ellishg) [reply lib/IRGen/IRGenSIL.cpp:2272]: Maybe I'm misunderstanding something but I'm already testing this line in IRGen, `markUsed(Bar.directClassMethod())`.
https://github.com/apple/swift/pull/33349/files#diff-836f9a570b4e9833ab0598773438d6dbR39

File: test/IRGen/objc_direct.swift
Comment (ellishg) [reply test/IRGen/objc_direct.swift]: By the way, I've added test for this in `test/type/objc_direct.swift`.

File: lib/IRGen/IRGenSIL.cpp
Comment (rjmccall) [reply lib/IRGen/IRGenSIL.cpp:2272]: You're not testing that we force the class to be realized before we make the call.

Comment (ellishg) [reply lib/IRGen/IRGenSIL.cpp:2272]: Ok, but I don't understand how to realize a class. Still have a lot to learn about Swift/Objective-C

Comment (rjmccall) [reply lib/IRGen/IRGenSIL.cpp:2475]: I just mean that you should have an IRGen test that checks for the exact sequence of materialization for the class argument, and you should leave a comment saying that we need to continue to do this for direct calls even if we stop doing it normally.

File: lib/SIL/IR/SILDeclRef.cpp
Comment (rjmccall) [reply lib/SIL/IR/SILDeclRef.cpp]: This has been committed to trunk; you might need to shepherd it into the stable branch.

File: lib/IRGen/IRGenSIL.cpp
Comment (ellishg) [reply lib/IRGen/IRGenSIL.cpp:2475]: Ah thanks, I finally understand! I've added more checks here:
https://github.com/apple/swift/pull/33349/files#diff-5115fb955c7cba0a551223ce515367063ba8f83c939474d54515f4477bd4efe4R39-R51

Comment (rjmccall) [reply lib/IRGen/IRGenSIL.cpp:2475]: Thanks, that's exactly what I wanted.

**Review by rjmccall — APPROVED** [review]
LGTM

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by ellishg** [comment]
Hopefully I have fixed the ci tests

**Comment by ellishg** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
https://github.com/apple/llvm-project/pull/2028

@swift-ci Please test

**Comment by ellishg** [comment]
Sorry, I think I introduced some errors when I fixed the merge conflict. I’ll take a look tomorrow.

File: test/Index/Inputs/cross_language.m
Comment (ellishg) [inline test/Index/Inputs/cross_language.m:9]: @rjmccall I want to draw your attention to this test I fixed that breaks when using apple/llvm-project#2028.

**Comment by ellishg** [comment]
I believe I have now fixed all the test failures.

**Comment by rjmccall** [comment]
@swift-ci Please test

File: test/Index/Inputs/cross_language.m
Comment (rjmccall) [reply test/Index/Inputs/cross_language.m:9]: Thanks

**Comment by ellishg** [comment]
@rjmccall Oh yeah. You'll need to test with apple/llvm-project#2028.

**Comment by rjmccall** [comment]
apple/llvm-project#2028

@swift-ci Please test

**Comment by rjmccall** [comment]
> @rjmccall Oh yeah. You'll need to test with [apple/llvm-project#2028](https://github.com/apple/llvm-project/pull/2028).

Argh, yes, I did it right earlier.

**Comment by ellishg** [comment]
@rjmccall The CI is failing to checkout `llvm-project`
> fatal: Couldn't find remote ref pull/2028/merge

Shall we try again?

**Comment by rjmccall** [comment]
> @rjmccall The CI is failing to checkout `llvm-project`
> 
> > fatal: Couldn't find remote ref pull/2028/merge
> 
> Shall we try again?

I'm just waiting until it merges into main; it should already be there, but the automerger is held up.

**Comment by rjmccall** [comment]
@swift-ci Please test

**Comment by rjmccall** [comment]
Alright, let's do it.  Thanks for seeing this through!

**Pull request merged**
